### PR TITLE
tests: introduce UI tests for core with diffs

### PR DESCRIPTION
**Pull request by tsusanka**
Replacement for #755.

I still have to improve the documentation, which I will do in the meantime. This adds the UI tests as discussed many times. It adds an argument for pytest called `ui` with two options:

- **record**: Create screenshots and calculate theirs hash. The screenshots are gitignored. The hash is included in git.
- **test**: Create screenshots, calculate theirs hash and test the hash against the one stored in git.

----

On CI:

- On success, all recorded screenshots are uploaded to https://firmware.corp.sldev.cz/ui_tests/ with a filename equal to the hash.

- On failure, the previously recorded screenshots are downloaded from https://firmware.corp.sldev.cz/ui_tests/ to create a diff.html file, which will inform you where the mismatch occurred. The Diff file will be stored in `tests/ui_tests/fixtures/[test_name]/diff.html`. From CI you can download it as an artifact in the `core unix device ui test` CI Job.

Diff.html example:

Screenshot_20191230_114736

----

Also this introduces a new env variable `TREZOR_DISABLE_ANIMATION` which replaces `TREZOR_DISABLE_FADE` and disables animation. It is required when running the UI tests.

----

Note:
- If you run the tests locally, they fail and you are not on VPN, they hang for a few seconds (the http download timeout). Are we okay with that? Isn't that confusing?
- <s>Should we have a seperate CI job for this or include it in the current `core unix device test`?</s> Agreed on no
- <s>This needs more attention: https://github.com/trezor/trezor-firmware/pull/784/files#diff-9f7414700342058eee129030ac8f0d30R226</s> Fixed

### CODE REVIEW

**Comment by prusnak** [comment]
Can we try the following to update the screenshots so they don't include the red square in the top-right corner?

Change this:

``` python
    def debug_display_refresh() -> None:
        display.bar(Display.WIDTH - 8, 0, 8, 8, 0xF800)
        display.refresh()
        if utils.SAVE_SCREEN:
            display.save("refresh")
```

into this:

``` python
    def debug_display_refresh() -> None:
        if utils.SAVE_SCREEN:
            display.save("refresh")
        display.bar(Display.WIDTH - 8, 0, 8, 8, 0xF800)
        display.refresh()
```

File: tests/conftest.py [hunk 1]
```diff
@@ -128,10 +128,25 @@ def client(request):
...
+
+    with ui_tests.screen_recording(client, request):
+        yield client
+
     client.close()
 
 
+def pytest_addoption(parser):
+    parser.addoption(
+        "--test_screen",
```
Comment (matejcik) [inline tests/conftest.py]: `--test-screen` please :)

File: core/src/trezor/ui/loader.py [hunk 2]
```diff
@@ -74,13 +74,13 @@ def on_render(self) -> None:
         else:
             s = self.active_style
 
-        Y = const(-24)
```
Comment (matejcik) [inline core/src/trezor/ui/loader.py]: is this change intentional?

**Comment by tsusanka** [comment]
That's a good idea! The code you posted does not work though, because the refresh function is called repeatedly, so you make a screenshot without the red corner then draw it and then you make a screenshot again, so you will end up with everything doubled.

We could redraw the corner with the previous color before doing the screenshot but we would have to add a function to `display` to get the color.  I'll think about it some more

File: core/src/trezor/ui/scroll.py [hunk 3]
```diff
@@ -32,6 +32,11 @@ def render_scrollbar(pages: int, page: int) -> None:
 
 
 def render_swipe_icon() -> None:
+    if utils.DISABLE_ANIMATION:
```
Comment (matejcik) [inline core/src/trezor/ui/scroll.py]: istm it would be nicer to limit the `DISABLE_ANIMATION` code just to set the value of `c` to a constant

File: tests/ui_tests/__init__.py [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+def _rename_records(screen_path):
+    # rename screenshots
+    for index, record in enumerate(sorted(screen_path.iterdir())):
+        filename = screen_path / "{:08}.png".format(index)
+        record.replace(filename)
+
+
+def _hash_files(files):
+    hasher = hashlib.sha256()
+    for file in sorted(files):
```
Comment (matejcik) [inline tests/ui_tests/__init__.py]:
```
for file in sorted(files):
    hasher.update(file.read_bytes())
```

**Comment by prusnak** [comment]
Or maybe we can just do the following:

```
    def debug_display_refresh() -> None:
        if utils.SAVE_SCREEN:
            display.save("refresh")
        else:
            display.bar(Display.WIDTH - 8, 0, 8, 8, 0xF800)
        display.refresh()
```

(i don't mind no showing the red dot in this special mode doing screenshots).

File: tests/ui_tests/__init__.py [hunk 5]
```diff
@@ -0,0 +1,128 @@
...
+
+
+@contextmanager
+def screen_recording(client, request):
+    if not request.node.get_closest_marker("skip_ui"):
+        test_screen = request.config.getoption("test_screen")
+    else:
+        test_screen = ""
+
+    if not test_screen:
```
Comment (matejcik) [inline tests/ui_tests/__init__.py]: instead:
```
test_screen = request.config.getoption("test_screen")
if request.node.get_closest_marker("skip_ui") or not test_screen:
    yield
    return
```

File: ci/test.yml [hunk 6]
```diff
@@ -38,6 +38,26 @@ core unix unit test:
...
+    - cd core
+    - pipenv run make test_emu_ui
+    - cp /var/tmp/trezor.log ${CI_PROJECT_DIR}
+    - cd ../ci
+    - pipenv run python prepare_ui_artifacts.py
+  artifacts:
+    name: core-unix-device-ui-test
+    paths:
+      - trezor.log
+      - ci/tmp/
```
Comment (matejcik) [inline ci/test.yml]: maybe a more distinct directory name for this?

File: core/src/trezor/ui/loader.py (hunk 2 above)
Comment (tsusanka) [reply core/src/trezor/ui/loader.py]: It's a rename to _Y. It was done by @jpochyla so I left it. You're right that I am not sure what was the rational behind this

File: tests/ui_tests/__init__.py (hunk 4 above)
Comment (tsusanka) [reply tests/ui_tests/__init__.py]: Fixed. To be pushed.

File: tests/conftest.py (hunk 1 above)
Comment (tsusanka) [reply tests/conftest.py]: Fixed. To be pushed.

File: tests/ui_tests/__init__.py [hunk 7]
```diff
@@ -0,0 +1,128 @@
...
+    return hasher.digest().hex()
+
+
+def _process_tested(fixture_test_path, test_name):
+    hash_file = fixture_test_path / "hash.txt"
+
+    if not hash_file.exists():
+        raise ValueError("File hash.txt not found.")
+
+    with open(hash_file, "r") as f:
```
[the remark this answers is not in the document]
Comment (tsusanka) [reply tests/ui_tests/__init__.py]: Fixed.

File: tests/ui_tests/__init__.py [hunk 8]
```diff
@@ -0,0 +1,128 @@
...
+    shutil.rmtree(screen_path, ignore_errors=True)
+    screen_path.mkdir()
+
+
+def _process_recorded(screen_path):
+    records = sorted(screen_path.iterdir())
+
+    # create hash
+    digest = _hash_files(records)
+    with open(screen_path / "../hash.txt", "w") as f:
```
[the remark this answers is not in the document]
Comment (tsusanka) [reply tests/ui_tests/__init__.py]: Fixed.

File: tests/ui_tests/__init__.py [hunk 9]
```diff
@@ -0,0 +1,128 @@
...
+
+def _get_test_dirname(node):
+    # This composes the dirname from the test module name and test item name.
+    # Test item name is usually function name, but when parametrization is used,
+    # parameters are also part of the name. Some functions have very long parameter
+    # names (tx hashes etc) that run out of maximum allowable filename length, so
+    # we limit the name to first 100 chars. This is not a problem with txhashes.
+    node_name = re.sub(r"\W+", "_", node.name)[:100]
+    node_module_name = node.getparent(pytest.Module).name
+    return "{}_{}".format(node_module_name, node_name)
```
[the remark this answers is not in the document]
Comment (tsusanka) [reply tests/ui_tests/__init__.py]: Fixed.

File: tests/ui_tests/__init__.py [hunk 10]
```diff
@@ -0,0 +1,128 @@
...
+    # delete old files
+    shutil.rmtree(screen_path, ignore_errors=True)
+    screen_path.mkdir()
+
+
+def _process_recorded(screen_path):
+    records = sorted(screen_path.iterdir())
+
+    # create hash
+    digest = _hash_files(records)
```
[the remark this answers is not in the document]
Comment (tsusanka) [reply tests/ui_tests/__init__.py]: Fixed. I agree.

File: tests/ui_tests/__init__.py [hunk 11]
```diff
@@ -0,0 +1,128 @@
...
+    with open(screen_path / "../hash.txt", "w") as f:
+        f.write(digest)
+    _rename_records(screen_path)
+
+
+def _rename_records(screen_path):
+    # rename screenshots
+    for index, record in enumerate(sorted(screen_path.iterdir())):
+        filename = screen_path / "{:08}.png".format(index)
```
[the remark this answers is not in the document]
Comment (tsusanka) [reply tests/ui_tests/__init__.py]: Fixed.

File: tests/ui_tests/__init__.py [hunk 12]
```diff
@@ -0,0 +1,128 @@
...
+        raise ValueError("File hash.txt not found.")
+
+    with open(hash_file, "r") as f:
+        expected_hash = f.read()
+
+    actual_path = fixture_test_path / "actual"
+    _rename_records(actual_path)
+
+    records = sorted(actual_path.iterdir())
+    actual_hash = _hash_files(records)
```
[the remark this answers is not in the document]
Comment (tsusanka) [reply tests/ui_tests/__init__.py]: Fixed.

File: tests/ui_tests/__init__.py [hunk 13]
```diff
@@ -0,0 +1,128 @@
...
+    else:
+        test_screen = ""
+
+    if not test_screen:
+        yield
+        return
+
+    fixture_root = Path(__file__) / "../fixtures"
+    test_name = _get_test_dirname(request.node)
+    fixture_test_path = fixture_root.resolve() / test_name
```
[the remark this answers is not in the document]
Comment (tsusanka) [reply tests/ui_tests/__init__.py]: Fixed.

File: ci/prepare_ui_artifacts.py [hunk 14]
```diff
@@ -0,0 +1,30 @@
...
+
+def _compare_hash(test_dir, hash):
+    with open(test_dir / "hash.txt", "r") as f:
+        content = f.read()
+        assert hash == content
+
+
+fixture_root = Path().cwd() / "../tests/ui_tests/fixtures"
+
+for test_dir in fixture_root.iterdir():
```
[the remark this answers is not in the document]
Comment (tsusanka) [reply ci/prepare_ui_artifacts.py]: Fixed.

File: ci/prepare_ui_artifacts.py [hunk 15]
```diff
@@ -0,0 +1,30 @@
+import hashlib
+import shutil
+from pathlib import Path
+
+
+def _hash_files(files):
```
[the remark this answers is not in the document]
Comment (tsusanka) [reply ci/prepare_ui_artifacts.py]: Fixed.

File: tests/ui_tests/html.py [hunk 16]
```diff
@@ -0,0 +1,68 @@
...
+            for r, a in zip_longest(recorded, actual):
+                if r and a and filecmp.cmp(a, r):
+                    background = "white"
+                else:
+                    background = "red"
+                with tr(bgcolor=background):
+                    _image(r)
+                    _image(a)
+
+    with open(fixture_test_path / "diff.html", "w") as f:
```
[the remark this answers is not in the document]
Comment (tsusanka) [reply tests/ui_tests/html.py]: Fixed.

File: tests/ui_tests/__init__.py (hunk 5 above)
Comment (tsusanka) [reply tests/ui_tests/__init__.py]: Fixed.

File: core/src/trezor/ui/loader.py (hunk 2 above)
Comment (jpochyla) [reply core/src/trezor/ui/loader.py]: `const()` without `_` creates a global even in the function scope

File: ci/test.yml (hunk 6 above)
Comment (tsusanka) [reply ci/test.yml]: Initial idea was we could use this for smth else as well. But it probably does not make sense. So `ui_test_fixtures` or `ui_test_records` ?

**Comment by tsusanka** [comment]
Great suggestions. I've applied most of it in 56257eb. Let's see what CI thinks. Would you like to confirm the changes by resoloving the conversation?

File: ci/test.yml (hunk 6 above)
Comment (matejcik) [reply ci/test.yml]: honestly whichever. `ui_test_records` seems a little nicer

File: core/src/trezor/ui/scroll.py (hunk 3 above)
Comment (tsusanka) [reply core/src/trezor/ui/scroll.py]: Fixed in https://github.com/trezor/trezor-firmware/pull/784/commits/47419c77688a66c5d32b06edbc8f0afbdb33b95f

File: ci/test.yml (hunk 6 above)
Comment (tsusanka) [reply ci/test.yml]: Fixed in 362dbce60

File: tests/device_tests/test_msg_change_wipe_code_t2.py [hunk 17]
```diff
@@ -143,6 +143,7 @@ def test_set_remove_wipe_code(client):
     assert client.features.wipe_code_protection is False
 
 
+@pytest.mark.setup_client()
```
Comment (matejcik) [inline tests/device_tests/test_msg_change_wipe_code_t2.py]: iirc you never need `setup_client` without arguments? the defaults will be used anyway

File: tests/ui_tests/report.py [hunk 18]
```diff
@@ -0,0 +1,104 @@
...
+from dominate.tags import div, h1, hr, i, img, p, table, td, th, tr
+
+from . import download
+
+
+def _image(src):
+    with td():
+        if src:
+            # open image file
+            image = open(src, "rb")
```
Comment (matejcik) [inline tests/ui_tests/report.py]: tiny nitpick but still: `with open(src, "rb") as f: ...` makes sure the image is closed afterwards

Comment (matejcik) [reply tests/ui_tests/report.py]: also as it turns out `src` is a Path, so instead `image = src.read_bytes()` :)

File: tests/conftest.py [hunk 19]
```diff
@@ -127,11 +137,53 @@ def client(request):
...
+    test_names = {"passed": set(), "failed": set()}
+    for status, test in reporter.stats.items():
+        if status in ("deselected", "warnings"):
+            continue
+        if status in ("passed", "failed"):
+            # iterate through the stages to get the test name
+            for t in test:
+                test_names[status].add(get_test_name(t.nodeid))
+
+    report.index(sorted(list(test_names["passed"])), sorted(list(test_names["failed"])))
```
Comment (matejcik) [inline tests/conftest.py]: istm you don't need `list()` here, and the `sorted()` should be moved either to `report.index` or maybe even inside `_report_links`

**Comment by matejcik** [comment]
:+1: ship it
